Fix Tile reference counting assertion with implicit tiling

kring opened the pull request:
@j9liu found a tileset that was causing reference count assertion errors when clicking "Refresh Tileset" in Cesium for Unreal. The problem essentially boiled down to the fact that the tileset used implicit tiling and some tiles that exist but do not have content, and as a result `ImplicitOctreeLoader` (also quadtree) creates tiles that initially have `TileEmptyContent`. Loaded content (including empty content) "sometimes" counts as a reference against the Tile. In this case, `addReference` was not being called when the tile was initially created with empty content, but `releaseReference` _was_ being called when the tile's content was unloaded. Oops!

This is more complicated than I'd like it to be. Loaded content counts as a `Tile` reference so that we don't destroy the tile before its content is unloaded. But we only want to do that for tile that _can_ be unloaded, right? And our rule is that we can only unload content if the tile has a valid ID, because without a valid ID, there's no possible way we can _reload_ the content if we unload it. Unloading content that we can't reload will break the tileset.

So, the overly-complicated rule is that content only counts as a reference against the `Tile` when the `Tile` also has an ID.

The `Tile` constructor has this code:

```cpp
if (this->hasReferencingContent()) {
  // Add a reference for the loaded content.
  this->addReference("Constructor with content");
}
```

Which looks like it should do the job.

Unfortunately, `hasReferencingContent` will _always_ return false, because at the time this code is invoke, the `Tile` can't possibly have an ID yet. So we're never adding a content reference.

In the vast majority of cases, this is fine. Most tiles with initial content will never get an ID anyway.

But in `ImplicitQuadtreeLoader` and `ImplicitOctreeLoader`, a `Tile` was constructed like this:

```cpp
if (subtreeAvailability.isContentAvailable(
                relativeChildLevel,
                relativeChildMortonID,
                0)) {
  children.emplace_back(&loader);
} else {
  children.emplace_back(&loader, TileEmptyContent{});
}

Tile& child = children.back();
// ...
child.setTileID(childID);
```

So, when an implicit tile doesn't have content available, we created it with `TileEmptyContent`, which will _not_ add a reference, as described above. And _then_ we give it an ID, which makes this `TileEmptyContent` unloadable, and when it's unloaded the reference count will be decremented. That's a `releaseReference` without an `addReference`, which eventually causes an assertion failure.

There are a number of possible solutions to this. The one I went with in this PR is not necessarily the best, but it's simple and low impact, and at least makes this sort of problem less likely to happen in the future.

I changed the `Tile` constructors that take content to _also_ require a `TileID`. With the `TileID` supplied at construction time, the constructor code will properly increment the reference count when it needs to do so.

It's still possible to subvert this system and run into reference counting issues by initially supplying a default `TileID` (empty string) and changing it later, but... don't do that. Really the `Tile` class could use some refactoring to better control its own internal state and make this sort of problem impossible. That's a much bigger change, though.

j9liu commented on CHANGES.md:
```diff
@@ -6,6 +6,7 @@
 
 - Renamed `SubtreeWriter::writeSubtree` to `SubtreeWriter::writeSubtreeJson`.
 - `SubtreeAvailability::createEmpty` now requires a boolean parameter to set initial tile availability.
+- `Cesium3DTilesSelection::Tile` constructors taking initial empty or external content now also require that a `TileID` be supplied.
```
Nitpick:

```suggestion
- `Cesium3DTilesSelection::Tile` constructors that take initially empty or external content now also require a `TileID` to be supplied.
```

j9liu commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h:
```diff
@@ -136,30 +136,56 @@ class CESIUM3DTILESSELECTION_API Tile final {
    * with this constructor.
    *
+   * If the supplied `TileID` is not an empty string, the tile's reference count
+   * will be incremented on account of the the loaded, referencable content.
```
```suggestion
   * will be incremented on account of the loaded, referencable content.
```

j9liu commented on Cesium3DTilesSelection/include/Cesium3DTilesSelection/Tile.h:
```diff
@@ -136,30 +136,56 @@ class CESIUM3DTILESSELECTION_API Tile final {
    * with this constructor.
    *
+   * If the supplied `TileID` is not an empty string, the tile's reference count
+   * will be incremented on account of the the loaded, referencable content.
```
```suggestion
   * will be incremented on account of the loaded, referencable content.
```

j9liu commented on Cesium3DTilesSelection/test/TestTilesetContentManager.cpp:
```diff
@@ -476,6 +476,7 @@ TEST_CASE("Test tile state machine") {
     pMockedLoader->mockCreateTileChildren = {{}, TileLoadResultState::Success};
```
Is it possible to design a unit test to catch this case? (Maybe in the implicit loader tests?)

j9liu reviewed (commented):
Thanks so much @kring ! I confirmed that this fixes the error I was seeing for that tileset. I just had a few comments, mostly just pointing out some typos 😋

kring commented:
Thanks for the review @j9liu! I fixed the typos (apparently I was aiming for a typo in every comment! 😆) and added some tests, so this should ready for another look.

j9liu commented on Cesium3DTilesSelection/test/TestImplicitOctreeLoader.cpp:
```diff
@@ -708,4 +708,68 @@ TEST_CASE("Test tile subdivision for implicit octree loader") {
+    }
+
+    // The seven tiles without content should have referencing EmptyContent.
+    // The fourth should not yet have any referencing content because it has not
```
```suggestion
    // The eighth should not yet have any referencing content because it has not
```
